idb/postgres/postgres_migrations.go

```diff
@@ -342,70 +382,296 @@ func m5accountCumulativeRewardsUpdate(db *IndexerDb, state *MigrationState) erro
+type createClose struct {
+	created sql.NullInt64
+	closed  sql.NullInt64
+}
+
+// updateClose will only allow the value to be set once.
+func updateClose(cc *createClose, value uint64) *createClose {
+	if cc == nil {
+		return &createClose{
+			closed: sql.NullInt64{
+				Valid: true,
+				Int64: int64(value),
+			},
+		}
+	}
+
+	if !cc.closed.Valid {
+		cc.closed.Valid = true
+		cc.closed.Int64 = int64(value)
+	}
+
+	return cc
+}
+
+// updateCreate will update the created round.
+func updateCreate(cc *createClose, value uint64) *createClose {
+	if cc == nil {
+		return &createClose{
+			created: sql.NullInt64{
+				Valid: true,
+				Int64: int64(value),
+			},
+		}
+	}
+
+	cc.created.Valid = true
+	cc.created.Int64 = int64(value)
+
+	return cc
+}
+
+func executeForEachCreatable(stmt *sql.Stmt, address []byte, m map[uint64]*createClose) (err error) {
+	for index, round := range m {
+		_, err = stmt.Exec(address, index, round.created, round.closed)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+type m5AccountData struct {
+	cumulativeRewards types.MicroAlgos
+	account           createClose
+	asset             map[uint64]*createClose
+	assetHolding      map[uint64]*createClose
+	app               map[uint64]*createClose
+	appLocal          map[uint64]*createClose
+
+}
+
+func initM5AccountData() *m5AccountData {
+	return &m5AccountData{
+		cumulativeRewards: 0,
+		account:           createClose{},
+		asset:             make(map[uint64]*createClose),
+		assetHolding:      make(map[uint64]*createClose),
+		app:               make(map[uint64]*createClose),
+		appLocal:          make(map[uint64]*createClose),
+	}
+}
+
+// processAccountTransactions contains all the accounting logic to recompute total rewards and create/close rounds.
+func processAccountTransactions(txnrows <-chan idb.TxnRow, addressStr string, address types.Address) (error, *m5AccountData) {
+	var err error
+	result := initM5AccountData()
+	numTxn := 0
+
+	// Loop through transactions
+	for txn := range txnrows {
+		if len(txn.TxnBytes) == 0 {
+			return fmt.Errorf("%s: processing account %s: found empty TxnBytes (rnd %d, intra %d):  %v", rewardsCreateCloseUpdateErr, addressStr, txn.Round, txn.Intra, err), nil
+			continue
+		}
+		numTxn++
+
+		// Transactions are ordered most recent to oldest, make sure created is set to the last transaction.
+		result.account.created.Valid = true
+		result.account.created.Int64 = int64(txn.Round)
+
+		// process transactions one at a time
+		var stxn types.SignedTxnWithAD
+		err = msgpack.Decode(txn.TxnBytes, &stxn)
+		if err != nil {
+			return fmt.Errorf("%s: processing account %s: decoding txn (rnd %d, intra %d):  %v", rewardsCreateCloseUpdateErr, addressStr, txn.Round, txn.Intra, err), nil
+		}
+
+		// When the account is closed rewards reset to zero.
+		// Because transactions are newest to oldest, stop accumulating once we see a close.
+		if !result.account.closed.Valid {
```

Comment:
This is true. I missed this edge case originally. There is a problem if I reset closed_at to null when things are re-created. So for now I'm skipping that. If something is re-created it will still have a "last_closed_at" field.